include/swift/SILOptimizer/Utils/InstOptUtils.h

```diff
@@ -587,6 +587,14 @@ FullApplySite cloneFullApplySiteReplacingCallee(FullApplySite applySite,
                                                 SILValue newCallee,
                                                 SILBuilderContext &builderCtx);
 
+/// Analyze the destructor for the class of \p value to see if any instructions
+/// in it could have side effects on the program outside the destructor.
+bool doesDestructorHaveSideEffects(SILValue value);
+
+/// Analyzing the body of this class destructor is valid because the object is
+/// dead. This means that the object is never passed to
+/// objc_setAssociatedObject, so its destructor cannot be extended at runtime.
+SILFunction *getDestructor(SILValue value);
```

Comment:
I'm not sure if it's a good idea to pass a SILValue instead of an AllocRefInst.
In case of anything else than an alloc_ref it's not guaranteed that the returned destructor is really the destructor which is called. It could also be the destructor of a derived class.

Reply:
That's a fair point. The use case I have in mind is getting the destructor of the operand of a release and passing it to `doesDestructorHaveSideEffects`. Maybe we could inline it into `doesDestructorHaveSideEffects` or add a comment that users should be careful? What do you think a good solution would be?

Reply:
What you can do is: check if it's an alloc_ref _or_ a final class and otherwise return null.

Reply:
Actually I'm wrong. Unfortunately there is this thing with associated objects. A class instance can have an associated object, which can have any kind of side effects on destruction.
So even a final class or if value is an alloc_ref, you cannot be sure about the side effects of the destructor.
It only works in dead object elimination, because if the object is dead, it's guaranteed that it's never passed to a call to setAssociatedObject.

Sorry, I didn't think of that. The conclusion is that generalizing this API is problematic because the user of doesDestructorHaveSideEffects must make sure that the value does not have an associated object.

Reply:
No worries. That makes sense. I'll close it.